### PR TITLE
Allow `try`-blocks in places where an open delim is expected

### LINKED ISSUE #56828: try_block cannot be used everywhere an expression can
The following code fails to compile (`rustc --edition=2018`):
```
#![feature(try_blocks)]
fn main() {
    match try { 0 } {
        None => (),
        Some(_) => ()
    };
}
```

The compiler error is:
```
error: expected expression, found reserved keyword `try`
 --> test.rs:3:11
  |
3 |     match try { 0 } {
  |     ----- ^^^ expected expression
  |     |
  |     while parsing this match expression

```
Adding parentheses or braces, it will compile fine:
```
    match ( try { 0 } ) { ... } //warning: unnecessary parentheses around `match` head expression
    match { try { 0 } } { ... }
```
The same issue happens with other constructs such as:
```
    if let Some(_) = try { 0 } { }
    for _ in try { 0 } { } //should fail because of "cannot infer type", not "expected expression"
```

### LINKED ISSUE #70490: unused_parens incorrectly lints on `if let Ok(x) = (try { ...`
[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=32dff2ee98cb5c7d41ed4969e6aa8e1c)
This produces a lint:
```rust
if let Ok(x) = (try { }: Result<(), ()>) {}
```
But removing the parentheses produces an error.

This bug was already mentioned in #42528, but that issue is lost in feature-request purgatory since it asks for the version without parentheses to be allowed.  I'm opening this issue to track the incorrect lint specifically.

(Edit: Fixed the example.)

This issue has been assigned to @lcnr via [this comment](https://github.com/rust-lang/rust/issues/70490#issuecomment-605617622).

### PR DESCRIPTION
**Pull request by lcnr**
Closes #70490
Closes #56828

r? @Centril

### CODE REVIEW

File: src/librustc_lint/unused.rs [hunk 1]
```diff
@@ -371,7 +371,7 @@ trait UnusedDelimLint {
     fn is_expr_delims_necessary(inner: &ast::Expr, followed_by_block: bool) -> bool {
         followed_by_block
             && match inner.kind {
-                ast::ExprKind::Ret(_) | ast::ExprKind::Break(..) => true,
+                ExprKind::Ret(_) | ExprKind::Break(..) | ExprKind::TryBlock(_) => true,
```
Comment (Centril) [inline src/librustc_lint/unused.rs:374]: This feels like a parser bug to me. We have:
```rust
    fn is_try_block(&self) -> bool {
        self.token.is_keyword(kw::Try) &&
        self.look_ahead(1, |t| *t == token::OpenDelim(token::Brace)) &&
        self.token.uninterpolated_span().rust_2018() &&
        // Prevent `while try {} {}`, `if try {} {} else {}`, etc.
        !self.restrictions.contains(Restrictions::NO_STRUCT_LITERAL)
    }
```
but we allow this for `unsafe {}`, `async {}`, etc.

cc @petrochenkov

**Comment by Dylan-DPC** [comment]
@lcnr this has been reviewed

**Comment by lcnr** [comment]
Waited for a comment by @petrochenkov on this, as this changes the purpose of this PR.

I changed the parser to allow try in scrutinee position/in places where open braces are expected.

```rust
// This now compiles
match try {} {
    Ok(()) | Err(()) => (),
}
```

**Comment by Centril** [comment]
Did you intend to include that submodule bump?

**Comment by lcnr** [comment]
I rarely do :laughing: thank you for catching this

**Comment by Centril** [comment]
cc @scottmcm, this fixes https://github.com/rust-lang/rust/issues/56828 which we agreed was a bug at the time.

@bors r+

**Pull request merged**